File: Java/121-Best Time to Buy and Sell Stock.java
//121. Best Time to Buy and Sell Stock

class Solution {
    public int maxProfit(int[] prices) {
        
    int buy = Integer.MAX_VALUE ,sell = 0, left = 0 , right = 0;
        
    while (right != prices.length){
        if(prices[left] < buy){
            buy = prices[left];
            
        } else if(prices[left] > prices[right]){
            buy = prices[right];
            left = right;
        }
        
        
        if(sell <= (prices[right] - buy)){
            sell = prices[right] - buy;
        }
        right ++; 
    }   
     
        return sell;
    }
}

//same 2 pointer but cleaner code 
class Solution {
    public int maxProfit(int[] prices) {
        
        int left = 0, maxProf = 0, right = 1;

        while(right < prices.length){
            if(prices[left] > prices[right]){
                left = right;
                
            }

            if(maxProf < prices[right] - prices[left]){
                maxProf = prices[right] - prices[left];
                
            }

            right++;
        } 

        return maxProf;

    }
}

File: Python3/121.best-time-to-buy-and-sell-stock.py
#T: O(n)
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        lowest = prices[0]
        res = 0

        for i in prices:
            if i < lowest:
                lowest = i
            res = max(res, i - lowest)
        return res
    
class Solution: #2ptr
    def maxProfit(self, prices: List[int]) -> int:

        l = 0
        r = 1
        profit = 0

        while(r < len(s)):
            if prices[l] < prices[r]:
                profit = max(profit, (prices[r] - prices[l]))
            else:
                l = r
            
            r+=1
        
        return profit

#T: O(n^2)

class Solution: #2ptr
    def maxProfit(self, prices: List[int]) -> int:
        profit = 0

        for i in range(len(prices)):
            for j in range(i, len(prices)):
                profit = max(profit, (prices[j] - prices[i]))

        return profit

       